src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2760,49 +2757,93 @@ static int getSurplusQueuedTaskCount() {
      * @return runState on exit
      */
     private long tryTerminate(boolean now, boolean enable) {
-        long e = runState, isShutdown;
-        if ((e & STOP) == 0L) {
-            if (now)
-                runState = e = (lockRunState() + RS_LOCK) | STOP | SHUTDOWN;
-            else if ((isShutdown = (e & SHUTDOWN)) != 0 || enable) {
-                if (isShutdown == 0)
-                    getAndBitwiseOrRunState(SHUTDOWN);
-                if (quiescent() > 0)
-                    e = runState;
-            }
-            if ((e & STOP) != 0L && (releaseAll() & RC_MASK) > 0L && now)
+        long e, isShutdown, ps;
+        if (((e = runState) & TERMINATED) != 0L)
+            now = false;
+        else if ((e & STOP) != 0L)
+            now = true;
+        else if (now) {
+            if (((ps = getAndBitwiseOrRunState(SHUTDOWN|STOP) & STOP)) == 0L) {
+                if ((ps & RS_LOCK) != 0L) {
+                    spinLockRunState(); // ensure queues array stable after stop
+                    unlockRunState();
+                }
                 interruptAll();
+            }
         }
-        if ((e & (STOP | TERMINATED)) == STOP) { // help cancel tasks
-            if ((ctl & RC_MASK) > 0L) {          // unless all inactive
-                int r = (int)Thread.currentThread().threadId();
-                WorkQueue[] qs = queues;         // stagger traversals
-                int n = (qs == null) ? 0 : qs.length;
-                for (int l = n; l > 0; --l, ++r) {
-                    WorkQueue q; ForkJoinTask<?> t;
-                    if ((q = qs[r & (n - 1)]) != null &&
-                        q.source != DROPPED) {
-                        while ((t = q.poll()) != null) {
-                            try {
-                                t.cancel(false);
-                            } catch (Throwable ignore) {
-                            }
-                        }
+        else if ((isShutdown = (e & SHUTDOWN)) != 0L || enable) {
+            if (isShutdown == 0L)
+                getAndBitwiseOrRunState(SHUTDOWN);
+            if (quiescent() > 0)
+                now = true;
+        }
+        if (now) {
```

Comment:
@DougLea I'd probably add a blank line above this one to signal that this if is not a part of the previous if-else chain.